crates/ruff_linter/src/rules/refurb/rules/slice_to_remove_prefix_or_suffix.rs

```diff
@@ -0,0 +1,498 @@
+    let ast::ExprSubscript { value, slice, .. } = value.as_subscript_expr()?;
+
+    affix_removal_data_core_logic(value, test, else_or_target_name, slice)
+}
+
+/// Suppose given a statement of the form:
+/// ```python
+/// if test:
+///     else_or_target_name = value[slice]
+/// ```
+/// or an expression of the form:
+/// ```python
+/// value[slice] if test else else_or_target_name
+/// ```
+/// This function verifies that
+///     - `value` and `else_or_target_name`
+/// are equal to a common name `text`
+///     - `test` is of the form `text.startswith(prefix)`
+/// or `text.endswith(suffix)`
+///     - `slice` has no upper bound in the case of a prefix,
+/// and no lower bound in the case of a suffix
+///
+/// If these conditions are satisfied, the function
+/// returns the corresponding `RemoveAffixData` object;
+/// otherwise it returns `None`.
+fn affix_removal_data_core_logic<'a>(
+    value: &'a ast::Expr,
+    test: &'a ast::Expr,
+    else_or_target_name: &'a ast::name::Name,
+    slice: &'a ast::Expr,
+) -> Option<RemoveAffixData<'a>> {
+    let body_name = &value.as_name_expr()?.id;
+    let slice = slice.as_slice_expr()?;
+    let test_name = &test
+        .as_call_expr()?
+        .func
+        .as_attribute_expr()?
+        .value
+        .as_name_expr()?
+        .id;
+    let func_name = test
+        .as_call_expr()?
+        .func
+        .as_attribute_expr()?
+        .attr
+        .id
+        .as_str();
+
+    let func_args = &test.as_call_expr()?.arguments.args;
+    if func_args.len() != 1 {
+        return None;
+    }
+
+    let affix = &func_args[0];
+    match (body_name, test_name, else_or_target_name, slice, func_name) {
+        (text, a, b, _, "startswith") if text == a && a == b && slice.upper.is_none() => {
+            Some(RemoveAffixData {
+                text,
+                bound: slice.lower.as_ref()?,
+                affix_query: AffixQuery {
+                    kind: AffixKind::StartsWith,
+                    affix,
+                },
+            })
+        }
+        (text, a, b, _, "endswith") if text == a && a == b && slice.lower.is_none() => {
+            Some(RemoveAffixData {
+                text,
+                bound: slice.upper.as_ref()?,
+                affix_query: AffixQuery {
+                    kind: AffixKind::EndsWith,
+                    affix,
+                },
+            })
+        }
+        _ => None,
+    }
+}
+
+/// Tests whether the slice of the given string actually removes the
+/// detected affix.
+///
+/// For example, in the situation
+///
+/// ```python
+///  text[:bound] if text.endswith(suffix) else text
+/// ```
+///
+/// this function verifies that `bound == -len(suffix)` in two cases:
+///     - `suffix` is a string literal and `bound` is a number literal
+///     - `suffix` is an expression and `bound` is
+///     exactly `-len(suffix)` (as AST nodes, prior to evaluation.)
+fn affix_matches_slice_bound(data: &RemoveAffixData) -> bool {
```

Comment:
should this be a method on `RemoveAffixData`, if the function takes no other arguments?

Reply:
This now takes the `checker` as argument so that I can use the `builtins` suggestion you made below - so I've left this as a function on its own.